fieldtracing/fieldtracing.cpp

```diff
@@ -0,0 +1,1701 @@
+                        perBGrid,
+                        dPerBGrid,
+                        technicalGrid,
+                        fieldTracingParameters.reconstructionCoefficientsCache,
+                        fsgridCell[0], fsgridCell[1], fsgridCell[2],
+                        x
+                     );
+                     no.parameters[ionosphereParameters::UPMAPPED_BX] = SBC::ionosphereGrid.dipoleField(x[0],x[1],x[2],X,0,X) + perB[0];
+                     no.parameters[ionosphereParameters::UPMAPPED_BY] = SBC::ionosphereGrid.dipoleField(x[0],x[1],x[2],Y,0,Y) + perB[1];
+                     no.parameters[ionosphereParameters::UPMAPPED_BZ] = SBC::ionosphereGrid.dipoleField(x[0],x[1],x[2],Z,0,Z) + perB[2];
+                     
+                     nodeNeedsContinuedTracing[n] = 0;
+                     nodeTracingCoordinates[n] = {0,0,0};
+                     break;
+                  }
+               } // while(true)
+            } // for
+         } // pragma omp parallel
+         
+         // Globally reduce whether any node still needs to be picked up and traced onwards
+         std::vector<int> sumNodeNeedsContinuedTracing(nodes.size(), 0);
+         std::vector<std::array<Real, 3>> sumNodeTracingCoordinates(nodes.size());
+         MPI_Allreduce(nodeNeedsContinuedTracing.data(), sumNodeNeedsContinuedTracing.data(), nodes.size(), MPI_INT, MPI_SUM, MPI_COMM_WORLD);
+         if(sizeof(Real) == sizeof(double)) {
+            MPI_Allreduce(nodeTracingCoordinates.data(), sumNodeTracingCoordinates.data(), 3*nodes.size(), MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+            MPI_Allreduce(nodeTracingStepSize.data(), reducedNodeTracingStepSize.data(), nodes.size(), MPI_DOUBLE, MPI_MAX, MPI_COMM_WORLD);
+         } else {
+            MPI_Allreduce(nodeTracingCoordinates.data(), sumNodeTracingCoordinates.data(), 3*nodes.size(), MPI_FLOAT, MPI_SUM, MPI_COMM_WORLD);
+            MPI_Allreduce(nodeTracingStepSize.data(), reducedNodeTracingStepSize.data(), nodes.size(), MPI_FLOAT, MPI_MAX, MPI_COMM_WORLD);
+         }
+         for(uint n=0; n<nodes.size(); n++) {
+            if(sumNodeNeedsContinuedTracing[n] > 0) {
+               anyNodeNeedsTracing=true;
+               nodeNeedsContinuedTracing[n] = 1;
+               
+               // Update that nodes' tracing coordinates
+               nodeTracingCoordinates[n][0] = sumNodeTracingCoordinates[n][0] / sumNodeNeedsContinuedTracing[n];
+               nodeTracingCoordinates[n][1] = sumNodeTracingCoordinates[n][1] / sumNodeNeedsContinuedTracing[n];
+               nodeTracingCoordinates[n][2] = sumNodeTracingCoordinates[n][2] / sumNodeNeedsContinuedTracing[n];
+            }
+            nodeTracingStepSize[n] = reducedNodeTracingStepSize[n];
+         }
+         
+      } while(anyNodeNeedsTracing);
+      
+      logFile << "fsgrid coupling traced in " << toto << " iterations of the tracing loop." << endl;
```

Comment:
This could get an "(ionosphere)" in front of it.

Reply:
ok